File: 287-FindTheDuplicateNumber/287-FindTheDuplicateNumber.java
// Last updated: 11/2/2025, 8:23:18 PM
class Solution {
    public int findDuplicate(int[] nums) {
    
        //LinkedList - Fast and Slow Pointer 
        int slow = 0;
        int fast = 0;

        do {
            slow = nums[slow];
            fast = nums[nums[fast]];
        }
        while (slow != fast);
        
        int slow2 = 0;
        do {
            slow = nums[slow];
            slow2 = nums[slow2];
        }
        while (slow != slow2);

        return slow2; 
        //HashSet 
        //Time complexity - O(n)
        //Space complexity - O(n)

        /*
        HashSet<Integer> dup = new HashSet<Integer>();

        for(int i:nums){
            if(dup.add(i) == false){
                return i;
            }
        }
        return 0;
        */

        //Brute Force - nested forloop
        //Time complexity - O(n^2)
        //Space complexity - O(1)
        /*
        for(int i = 0; i <nums.length-1;i++){
            for(int y = i+1; y <nums.length; y++){
                if(nums[i] == nums[y]){
                    return nums[i];
                }
            }
        }
        return 0;
        */
    }
}

File: 287-FindTheDuplicateNumber/287-FindTheDuplicateNumber.py
# Last updated: 11/2/2025, 8:23:16 PM
class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        #Floyd's Tortoise and Hare Algorithm, linked list cycle 
        slow = 0
        fast = 0

        #find the first intersection between fast and slow pointers 
        while True:
            slow = nums[slow]
            fast = nums[nums[fast]]
            if slow == fast:
                break

        #make a new pointer slow2 and find the intersection between slow and slow2 
        #that value is the duplicate value and the start of the linked list cycle
        slow2 = 0
        while True:
            slow = nums[slow]
            slow2 = nums[slow2]
            if slow == slow2:
                return slow

        #time complexity:O(n)
        #space complexity:O(1)
